### ISSUE TITLE
Error in readRDS(dest) : unknown input format

### ISSUE THREAD

**Issue by amarchin** [rstudio/packrat#379]
When I try to invoke `packrat::status()` or `packrat::snapshot()`I get the error 

> Error in readRDS(dest) : unknown input format

and nothing happens.

The packrat version is 0.4.8-1 and these are the information of `sessionInfo()`:

```
R version 3.4.0 (2017-04-21)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Sierra 10.12.4

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib

locale:
[1] it_IT.UTF-8/it_IT.UTF-8/it_IT.UTF-8/C/it_IT.UTF-8/it_IT.UTF-8

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] RPostgreSQL_0.4-1    DBI_0.6-1            testthat_1.0.2      
 [4] forecastHybrid_0.4.0 prophet_0.1.1        Rcpp_0.12.10        
 [7] timekit_0.2.0        hts_5.0              matrixcalc_1.0-3    
[10] Matrix_1.2-10        SparseM_1.77         forecast_8.0        
[13] xts_0.9-7            zoo_1.8-0            dplyr_0.5.0         
[16] purrr_0.2.2          readr_1.1.0          tidyr_0.6.1         
[19] tibble_1.3.0         ggplot2_2.2.1        tidyverse_1.1.1     
[22] lubridate_1.6.0      futile.logger_1.4.3  feather_0.3.1       
[25] doMC_1.3.4           iterators_1.0.8      foreach_1.4.3       
[28] config_0.2          

loaded via a namespace (and not attached):
 [1] lattice_0.20-35      assertthat_0.2.0     packrat_0.4.8-1     
 [4] lmtest_0.9-35        psych_1.7.3.21       R6_2.2.0            
 [7] cellranger_1.1.0     plyr_1.8.4           futile.options_1.0.0
[10] stats4_3.4.0         httr_1.2.1           lazyeval_0.2.0      
[13] readxl_1.0.0         fracdiff_1.4-2       stringr_1.2.0       
[16] foreign_0.8-68       munsell_0.4.3        broom_0.4.2         
[19] compiler_3.4.0       modelr_0.1.0         rstan_2.15.1        
[22] mnormt_1.5-5         nnet_7.3-12          gridExtra_2.2.1     
[25] quadprog_1.5-5       codetools_0.2-15     crayon_1.3.2        
[28] grid_3.4.0           nlme_3.1-131         jsonlite_1.4        
[31] gtable_0.2.0         magrittr_1.5         StanHeaders_2.15.0-1
[34] scales_0.4.1         stringi_1.1.5        reshape2_1.4.2      
[37] tseries_0.10-40      timeDate_3012.100    xml2_1.1.1          
[40] lambda.r_1.1.9       tools_3.4.0          forcats_0.2.0       
[43] hms_0.3              yaml_2.1.14          inline_0.3.14       
[46] colorspace_1.3-2     extraDistr_1.8.5     rvest_0.3.2         
[49] haven_1.0.0         
```

**Comment by amarchin** [comment]
The problem can be solved removing all the folders in the packrat folder (lib, lib-R, lib-ext, src) and then reinstalling the dependencies thanks to packrat.lock file. Anyway I don't like this kind of solution.

**Comment by peteratemarsys** [comment]
@amarchin I have a similar issue.

Unfortunately I can't replicate the issue from scratch but it does come back from time to time.

**what I experience:**
First, I add our local drat repo path to options. Then, after installing files from this local repo, and then running some of packrat's functions (such as snapshot/restore/status) sometimes, the following (same as above) error message appear:

> Error in readRDS(dest) : unknown input format

If the error once appear I can only make it go away by removing every packrat related files/folders (./packrat/*, .Rprofile) and re-initiating packrat within the project which I don't consider a solution. Also, it takes a considerably long time.



**What I found so far: **

The issue comes from available.packages: when we iterate through the repos (`for (repos in contriburl) { [...]`) contriburl is generated as if our local drat repos would contain binary packages (i.e. `"http://local-shared-folder/dratrepo/bin/macosx/el-capitan/contrib/3.4"`), however it should look for `"http://local-shared-folder/dratrepo/"` or `http://local-shared-folder/dratrepo/src/contrib/`.
Due to this error when we get to the point of downloading from the URL (i.e. `download.file(url = paste0(repos, "/PACKAGES.rds"), destfile = dest, method = method, cacheOK = FALSE, quiet = TRUE, mode = "wb")`) it will save an HTML syntax script that says: *404 URL not found* into the `dest` Rds file. So when we get to the `readRDS(dest)` or `readRDS(tmpf)` (I'm not sure which) part it will return the *unknown input format* error because it finds the HTML content instead of a proper .Rmd.

This suggests that when readRDS is runs to the error it's not contained in a `tryCatch` snippet.

Also, I'm not sure it's expected that our drat repo doesn't contain *file* in its name, so that `localcran <- length(grep("^file:", repos)) > 0L` is evaluated to `FALSE`.

I hope this helps to wrap this issue up and I'm looking forward to a solution to this. Please let me know if I could provide additional information.

**Comment by peteratemarsys** [comment]
One more note: setting the `pkgType` option to `source` (i.e. `options('pkgType' = 'source)`) seem to "solve" the issue. The reason it's not really a fix is that now even CRAN repos download from source (instead of binary).

**Comment by kevinushey** [comment]
This sounds like a feature request I made a while back:

https://github.com/eddelbuettel/drat/issues/46

It seems like either this is something that could be resolved in `drat`, or worked around by setting `options(pkgType = "source")` (which isn't ideal for the reasons you already enumerated)

**Comment by peteratemarsys** [comment]
Indeed! Great initiative and thanks for coming back on this. Do you think you could still do the PR?

**Comment by lc822** [comment]
Hi, 

I keep having this issue with packrat.

`Error in readRDS(dest) : unknown input format`

What is the current best approach to solve this?

Best wishes,
Lucy

**Comment by peteratemarsys** [comment]
Hi @lc822 ,

Sorry, our team didn't find a solution for this since then (except the above mentioned work around of setting the `options(pkgType = "source")`).
